src/hotp/hotp.c

```diff
@@ -493,11 +493,13 @@ uint8_t* page = (uint8_t *) current_slot_address;
 
     memcpy (page_buffer, page, SLOT_PAGE_SIZE);
 
-    // check if the secret from the tool is empty and if it is use the old
-    // secret
-uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
-
-    if (secret[0] == 0)
+    // Check if the secret from the tool is empty and if it is use the old secret
+    // Secret could begin with 0x00, so checking the whole secret before keeping the old one in mandatory
+    // The while last loop points secret[-1]. Well, it's right, == result does not matter and mem exits, we need a -1
+    uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
+    int8_t i = 20;
+    while (i >= 0 && secret[--i] == 0x00);
```

Comment:
Maybe I got confused with the post/pre-decrement again (whose use as value should be avoided imo, for this reason). Where did I go wrong:
When i is 0, the "i >= 0" is true, which means the right part of the condition is evaluated. Then "--i" evaluates to -1, which makes "secret[i]" secret[-1].

The following program prints "access to secret[-1]" for me.
```
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>

int main()
{
    uint8_t* secret = malloc(2)+1;
    secret[-1]=99;
    secret[0]=0;

    int8_t i = 0;
    if (i >= 0 && secret[--i] == 99) {
        printf("access to secret[-1]!\n");
    }
    return 0;
}
```